crates/exex/exex/src/backfill.rs

```diff
@@ -0,0 +1,355 @@
+};
+use reth_prune_types::PruneModes;
+use reth_revm::database::StateProviderDatabase;
+use reth_stages_api::{format_gas_throughput, ExecutionStageThresholds};
+use reth_tracing::tracing::{debug, trace};
+use std::{
+    marker::PhantomData,
+    ops::RangeInclusive,
+    time::{Duration, Instant},
+};
+
+/// Factory for creating new backfill jobs.
+#[derive(Debug, Clone)]
+pub struct BackfillJobFactory<E, P> {
+    executor: E,
+    provider: P,
+    prune_modes: PruneModes,
+    thresholds: ExecutionStageThresholds,
+}
+
+impl<E: Clone, P: Clone> BackfillJobFactory<E, P> {
+    /// Creates a new [`BackfillJobFactory`].
+    pub const fn new(
+        executor: E,
+        provider: P,
+        prune_modes: PruneModes,
+        thresholds: ExecutionStageThresholds,
+    ) -> Self {
+        Self { executor, provider, prune_modes, thresholds }
+    }
+
+    /// Creates a new backfill job for the given range.
+    pub fn backfill<DB>(&self, range: RangeInclusive<BlockNumber>) -> BackfillJob<E, DB, P> {
+        BackfillJob {
+            executor: self.executor.clone(),
+            provider: self.provider.clone(),
+            prune_modes: self.prune_modes.clone(),
+            range,
+            thresholds: self.thresholds.clone(),
+            _db: PhantomData,
+        }
+    }
+}
+
+impl BackfillJobFactory<(), ()> {
+    /// Creates a new [`BackfillJobFactory`] from [`FullNodeComponents`].
+    pub fn new_from_components<Node: FullNodeComponents>(
+        components: Node,
+        prune_modes: PruneModes,
+        thresholds: ExecutionStageThresholds,
+    ) -> BackfillJobFactory<Node::Executor, Node::Provider> {
+        BackfillJobFactory::<_, _>::new(
+            components.block_executor().clone(),
+            components.provider().clone(),
+            prune_modes,
+            thresholds,
+        )
+    }
+}
+
+/// Backfill job started for a specific range.
+///
+/// It implements [`Iterator`] that executes blocks in batches according to the provided thresholds
+/// and yields [`Chain`]
+#[derive(Debug)]
+pub struct BackfillJob<E, DB, P> {
+    executor: E,
+    provider: P,
+    prune_modes: PruneModes,
+    range: RangeInclusive<BlockNumber>,
+    thresholds: ExecutionStageThresholds,
+    _db: PhantomData<DB>,
+}
+
+impl<E, DB, P> Iterator for BackfillJob<E, DB, P>
+where
+    E: BlockExecutorProvider,
+    DB: Database,
+    P: FullProvider<DB>,
+{
+    type Item = Result<Chain, BlockExecutionError>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.range.is_empty() {
+            return None
+        }
+
+        Some(self.execute_range())
+    }
+}
+
+impl<E, DB, P> BackfillJob<E, DB, P>
+where
+    E: BlockExecutorProvider,
+    DB: Database,
+    P: FullProvider<DB>,
+{
+    fn execute_range(&mut self) -> Result<Chain, BlockExecutionError> {
+        let provider_ro =
+            self.provider.database_provider_ro()?.disable_long_read_transaction_safety();
```

Comment:
we should revisit some of the traits, because requiring the DB type is a bit stupid here...